### PR TITLE
Warn if the cooling efficiency is below 50%

### PR DESCRIPTION
**Pull request by nickshanks**
# Enhancement

## Summary
Adds one more preflight warning for when you have too many outfit expansions.

## Screenshots
Before:
Screenshot 2024-06-08 at 15 51 44
After:
Screenshot 2024-06-08 at 15 51 51
Subsequent:
Screenshot 2024-06-08 at 15 51 58

## Testing Done
Minimal, enough to get he screenshots. I didn't try to take off.

## Save File
Get a cargo vessel. Buy a lot of outfit expansions to test.
Add a lot of cooling. It should have near-zero effect for put an additional Liquid Helium Cooler once you have enough Outfit Expansions.
The idea is that you can remove both the expansions and the unnecessary cooling and improve the kit-out of the ship.

## Wiki Update
Don't think it needs one.

## Performance Impact
N/A

### CODE REVIEW

File: data/_ui/tooltips.txt [hunk 1]
```diff
@@ -1188,6 +1188,9 @@ tip "no fuel?"
 tip "no bays?"
 	`There are insufficient bays available to carry this ship. If it took off, it would be unable to leave this star system.`
 
+tip "inefficient cooling?"
+	`This ship has inefficient cooling. Remove some outfit expansions to improve efficiency.`
```
Comment (Amazinite) [inline data/_ui/tooltips.txt:1192]: I wouldn't want the tooltip to refer specifically to outfit expansions. While those are the only use of cooling inefficiency in vanilla at the moment, that might not always be the case, and there could always be plugins that make use of the cooling inefficiency attribute.

Comment (nickshanks) [reply data/_ui/tooltips.txt:1192]: I agree, but felt for now this was good enough. I *do* want to inform novice players about how to fix the situation, and at present, that means removing some Outfit Expansions.

My plan was that plug-ins could override the tool-tip text too.

**Review by TheGiraffe3 — APPROVED** [review]
This will be really nice to have. Thank you!

File: data/_ui/tooltips.txt (hunk 1 above)
Comment (TomGoodIdea) [reply data/_ui/tooltips.txt:1192]: In-game description of the OE says that "Unfortunately, rearranging the interior of your ship in such a fundamental way tends to interfere with its cooling systems", so the player should know that it can cause problems with cooling.
Maybe a way to filter the outfitter to show only items with a selected attribute would help locating the problematic outfits.

Comment (nickshanks) [reply data/_ui/tooltips.txt:1192]: "Can cause problems" is mild and doesn't imply the scaling function that CoolingEfficiency() applies. The scaling means that you need more than 15 tons of cooling per expansion added. I felt that 0.5 was a good value between "you're going the wrong way" and the red alert that already exists.
Re. cooling UI: Perhaps, and once we have that we can reword the tooltip.

Comment (nickshanks) [reply data/_ui/tooltips.txt]: > I wouldn't want the tooltip to refer specifically to outfit expansions. While those are the only use of cooling inefficiency in vanilla at the moment, that might not always be the case, and there could always be plugins that make use of the cooling inefficiency attribute.

Is this better? (New commit)

**Comment by Quantumshark** [comment]
This is somewhat of an arbitrary cutoff, and doesn't necessarily correspond to any particularly disastrous consequences. Certainly, not to any consequences that aren't visible on the heat table. I'm not sure it's an appropriate circumstance for a warning.

**Review by OcelotWalrus — APPROVED** [review]

**Comment by nickshanks** [comment]
> doesn't necessarily correspond to any particularly disastrous consequences

Is that not what warnings are for? The goal is to educate new players at precisely the point where they are making what is probably a mistake, and to be able to spot problem ships you may already have.

I chose 0.5 somewhat arbitrarily, because beyond that point you are adding more cooling than outfits in order to keep heat levels under control.

**Comment by warp-core** [comment]
The warning seems to say something like:
"You have [a considerable amount] of this undesirable attribute. Consider removing some of the attributes that add it."

Currently, it doesn't even check if you have any cooling at all. "This cooling you don't have is going to be very inefficient. Maybe you should remove some outfits that add cooling inefficiency to make your non-existent cooling more efficient."

All the other warnings relate to situations where the ship may not behave as a normal ship might. This warning is simply "you could optimise your outfits more". And, there's not actually any guarantee that this is true.

The original idea here seems to have been to convey that, with cooling outfits and outfits expansions, there may come a point where having less cooling and fewer outfits expansions is better than installing an outfits expansion to make space for more cooling outfits,
If this is something we need to explain to players, I think it should be done through the description of the outfits expansions.

**Comment by nickshanks** [comment]
> The original idea here seems to have been to convey that, with cooling outfits and outfits expansions, there may come a point where having less cooling and fewer outfits expansions is better than installing an outfits expansion to make space for more cooling outfits,

Yep, that was my intent.
Can you think of something better, and perhaps more general that deals with all negative effect properties.

> All the other warnings relate to situations where the ship may not behave as a normal ship might.

Having your ship overheat does cause it to not behave normally. I think that is a valid case for a warning.

**Comment by Quantumshark** [comment]
> Having your ship overheat does cause it to not behave normally. I think that is a valid case for a warning.

But this warning doesn't really connect with that. It's possible to have a ship that can overheat but doesn't get this warning, and it's possible to have a ship that won't overheat but does get this warning. As a metric for whether the ship overheats, this doesn't really work.

We already have a flight check failure for when the ship would overheat when idling. We _could_ add a warning for when the ship is at risk of overheating (i.e. `net change` on the heat chart exceeds `max` on the heat chart), but that's arguably redundant with those sections of the heat chart.

**Comment by nickshanks** [comment]
> We already have a flight check failure for when the ship would overheat when idling. We _could_ add a warning for when the ship is at risk of overheating (i.e. `net change` on the heat chart exceeds `max` on the heat chart), but that's arguably redundant with those sections of the heat chart.

It's also a common setup for ships (including game-defined fleets) so I would not want that. The goal isn't to warn the player their ship is at risk of overheating, but to warn them they are wasting cargo space adding an outfits expansion and 15 cubic metres of useless cooling to fix the inefficiencies it created.

**Comment by Quantumshark** [comment]
If the check is supposed to be "you'd be better using fewer outfits expansions and installing less cooling", then just looking for 50% cooling inefficiency doesn't work.

You'd need to check whether
current total cooling * proportional loss of cooling from installing the most recent outfits expansion
Is more than
amount of cooling in 15 tons * total cooling inefficiency

There are several problems with this:
1: It still assumes that outfits expansions are the only source of cooling inefficiency that exists, which won't necessarily be the case in future and/or with plugins.
Your whole premise that the player is "adding an outfits expansion and 15 cubic metres of useless cooling to fix the inefficiencies it created" only works if you assume that outfits expansions are the only source of cooling inefficiency that will ever exist.
2: `amount of cooling in 15 tons` isn't easy to work out either - if you've got 9 tons of space left over, adding one outfits expansion lets you fit a 24-ton large heat shunt, for 3240 cooling. Alternatively, three cooling ducts will only let you fit 900 cooling into 15 space, if they're somehow the only source of cooling that you have access to.

In short, the metric used here is a bad proxy for the situation you're trying to flag up, and I don't think there's _any_ way to accurately identify it.

**Comment by nickshanks** [comment]
I know there's no perfect metric, but I think this is reasonable, and would have helped me out a week ago before I understood what cooling inefficiency was. I have also removed all reference to outfit expansions.

Can you suggest an alternative that helps players who are unaware of the problem they have?

**Comment by warp-core** [comment]
> I know there's no perfect metric, but I think this is reasonable

I don't think this is reasonable.

As I mentioned in a previous comment, expanding the description of Outfits Expansions to provide more insight into how they can interfere with a ships cooling.

A warning that says the ship might be outfitted sub-optimally, which may not even be true, is not appropriate.
